File: typescript/triangle/triangle.test.ts
import { Triangle } from './triangle'

describe('Triangle', () => {
  describe('equilateral triangle', () => {
    it('all sides are equal', () => {
      const triangle = new Triangle(2, 2, 2)
      expect(triangle.isEquilateral).toBe(true)
    })

    it('any side is unequal', () => {
      const triangle = new Triangle(2, 3, 2)
      expect(triangle.isEquilateral).toBe(false)
    })

    it('no sides are equal', () => {
      const triangle = new Triangle(5, 4, 6)
      expect(triangle.isEquilateral).toBe(false)
    })

    it('all zero sides is not a triangle', () => {
      const triangle = new Triangle(0, 0, 0)
      expect(triangle.isEquilateral).toBe(false)
    })

    it('sides may be floats', () => {
      const triangle = new Triangle(0.5, 0.5, 0.5)
      expect(triangle.isEquilateral).toBe(true)
    })
  })

  describe('isosceles triangle', () => {
    it('last two sides are equal', () => {
      const triangle = new Triangle(3, 4, 4)
      expect(triangle.isIsosceles).toBe(true)
    })

    it('first two sides are equal', () => {
      const triangle = new Triangle(4, 4, 3)
      expect(triangle.isIsosceles).toBe(true)
    })

    it('first and last sides are equal', () => {
      const triangle = new Triangle(4, 3, 4)
      expect(triangle.isIsosceles).toBe(true)
    })

    it('equilateral triangles are also isosceles', () => {
      const triangle = new Triangle(4, 4, 4)
      expect(triangle.isIsosceles).toBe(true)
    })

    it('no sides are equal', () => {
      const triangle = new Triangle(2, 3, 4)
      expect(triangle.isIsosceles).toBe(false)
    })

    it('first triangle inequality violation', () => {
      const triangle = new Triangle(1, 1, 3)
      expect(triangle.isIsosceles).toBe(false)
    })

    it('second triangle inequality violation', () => {
      const triangle = new Triangle(1, 3, 1)
      expect(triangle.isIsosceles).toBe(false)
    })

    it('third triangle inequality violation', () => {
      const triangle = new Triangle(3, 1, 1)
      expect(triangle.isIsosceles).toBe(false)
    })

    it('sides may be floats', () => {
      const triangle = new Triangle(0.5, 0.4, 0.5)
      expect(triangle.isIsosceles).toBe(true)
    })
  })

  describe('scalene triangle', () => {
    it('no sides are equal', () => {
      const triangle = new Triangle(5, 4, 6)
      expect(triangle.isScalene).toBe(true)
    })

    it('all sides are equal', () => {
      const triangle = new Triangle(4, 4, 4)
      expect(triangle.isScalene).toBe(false)
    })

    it('two sides are equal', () => {
      const triangle = new Triangle(4, 4, 3)
      expect(triangle.isScalene).toBe(false)
    })

    it('may not violate triangle inequality', () => {
      const triangle = new Triangle(7, 3, 2)
      expect(triangle.isScalene).toBe(false)
    })

    it('sides may be floats', () => {
      const triangle = new Triangle(0.5, 0.4, 0.6)
      expect(triangle.isScalene).toBe(true)
    })
  })
})

File: dart/triangle/test/triangle_test.dart
import 'package:test/test.dart';
import 'package:triangle/triangle.dart';

final triangle = Triangle();

void main() {
  group('Triangle: equilateral triangle - ', equilateralTriangle);
  group('Triangle: isosceles triangle - ', isoscelesTriangle);
  group('Triangle: scalene triangle - ', scaleneTriangle);
}

void equilateralTriangle() {
  test('all sides are equal', () {
    final result = triangle.equilateral(2, 2, 2);
    expect(result, equals(true));
  }, skip: false);

  test('any side is unequal', () {
    final result = triangle.equilateral(2, 3, 2);
    expect(result, equals(false));
  }, skip: false);

  test('no sides are equal', () {
    final result = triangle.equilateral(5, 4, 6);
    expect(result, equals(false));
  }, skip: false);

  test('all zero sides is not a triangle', () {
    final result = triangle.equilateral(0, 0, 0);
    expect(result, equals(false));
  }, skip: false);

  test('sides may be floats', () {
    final result = triangle.equilateral(0.5, 0.5, 0.5);
    expect(result, equals(true));
  }, skip: false);
}

void isoscelesTriangle() {
  test('last two sides are equal', () {
    final result = triangle.isosceles(3, 4, 4);
    expect(result, equals(true));
  }, skip: false);

  test('first two sides are equal', () {
    final result = triangle.isosceles(4, 4, 3);
    expect(result, equals(true));
  }, skip: false);

  test('first and last sides are equal', () {
    final result = triangle.isosceles(4, 3, 4);
    expect(result, equals(true));
  }, skip: false);

  test('equilateral triangles are also isosceles', () {
    final result = triangle.isosceles(4, 4, 4);
    expect(result, equals(true));
  }, skip: false);

  test('no sides are equal', () {
    final result = triangle.isosceles(2, 3, 4);
    expect(result, equals(false));
  }, skip: false);

  test('first triangle inequality violation', () {
    final result = triangle.isosceles(1, 1, 3);
    expect(result, equals(false));
  }, skip: false);

  test('second triangle inequality violation', () {
    final result = triangle.isosceles(1, 3, 1);
    expect(result, equals(false));
  }, skip: false);

  test('third triangle inequality violation', () {
    final result = triangle.isosceles(3, 1, 1);
    expect(result, equals(false));
  }, skip: false);

  test('sides may be floats', () {
    final result = triangle.isosceles(0.5, 0.4, 0.5);
    expect(result, equals(true));
  }, skip: false);
}

void scaleneTriangle() {
  test('no sides are equal', () {
    final result = triangle.scalene(5, 4, 6);
    expect(result, equals(true));
  }, skip: false);

  test('all sides are equal', () {
    final result = triangle.scalene(4, 4, 4);
    expect(result, equals(false));
  }, skip: false);

  test('first and second sides are equal', () {
    final result = triangle.scalene(4, 4, 3);
    expect(result, equals(false));
  }, skip: false);

  test('first and third sides are equal', () {
    final result = triangle.scalene(3, 4, 3);
    expect(result, equals(false));
  }, skip: false);

  test('second and third sides are equal', () {
    final result = triangle.scalene(4, 3, 3);
    expect(result, equals(false));
  }, skip: false);

  test('may not violate triangle inequality', () {
    final result = triangle.scalene(7, 3, 2);
    expect(result, equals(false));
  }, skip: false);

  test('sides may be floats', () {
    final result = triangle.scalene(0.5, 0.4, 0.6);
    expect(result, equals(true));
  }, skip: false);
}
